File: 0494-target-sum/0494-target-sum.java
class Solution {
    int[][] dp;
    int t;
    public int findTargetSumWays(int[] nums, int target) {
        dp= new int[nums.length+1][2001];
        for(int[] row : dp) Arrays.fill(row, -1);
        t=target;
        return subsetsum(nums, nums.length, 0);
    }
    
    int subsetsum(int[] nums, int lnt, int trgt){
        if(lnt==0){
            if(trgt==t) return 1;
            else return 0;
        }
        if(dp[lnt][trgt+1000]!=-1) return dp[lnt][trgt+1000];
        dp[lnt][trgt+1000] = subsetsum(nums, lnt-1, trgt+nums[lnt-1]) + subsetsum(nums, lnt-1, trgt-nums[lnt-1]);
        return dp[lnt][trgt+1000];
    }
}

// class Solution {
//     public int findTargetSumWays(int[] nums, int target) {
//         return subsetsum(nums, nums.length, 0, target);
//     }
//     int subsetsum(int[] nums, int lnt, int curr, int trgt){
//         if(lnt==0){
//             if(trgt==curr) return 1;
//             else return 0;
//         }
//         return subsetsum(nums, lnt-1, curr+nums[lnt-1]) + subsetsum(nums, lnt-1, curr-nums[lnt-1]);
//     }
// }

File: 0494-target-sum/0494-target-sum.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        sumt = sum(nums)
        if abs(target)>sumt:
            return 0
        cols, rows = 2*sumt + 1, len(nums)+1
        dp = [[0]*(cols) for _ in range(rows)]
        
        for l in range(rows):
            for t in range(cols):
                if l==0:
                    if t==sumt:
                        dp[l][t] = 1
                    else:
                        dp[l][t] = 0
                if t-nums[l-1]>=0:
                    dp[l][t] += dp[l-1][t-nums[l-1]]
                if t+nums[l-1]<cols:
                    dp[l][t] += dp[l-1][t+nums[l-1]]
        
        return dp[len(nums)][target+sumt]